pkg/kv/kvserver/replica_application_decoder.go

```diff
@@ -73,10 +73,72 @@ func (d *replicaDecoder) decode(ctx context.Context, ents []raftpb.Entry) error
 	return nil
 }
 
+// retrieveLocalProposalsV2 is used with useReproposalsV2, replacing a call
+// to retrieveLocalProposals. The V2 implementation is simpler because a log
+// entry that comes up for a local proposal can always consume that proposal
+// from the map because V2 never mutates the MaxLeaseIndex for the same proposal.
+// In contrast, with V1, we can only remove the proposal from the map once we
+// have found a log entry that had a matching MaxLeaseIndex. This lead to the
+// complexity of having multiple entries associated to the same proposal during
+// application.
+func (d *replicaDecoder) retrieveLocalProposalsV2() (anyLocal bool) {
+	d.r.mu.Lock()
+	defer d.r.mu.Unlock()
+
+	var it replicatedCmdBufSlice
+
+	for it.init(&d.cmdBuf); it.Valid(); it.Next() {
+		cmd := it.cur()
+		cmd.proposal = d.r.mu.proposals[cmd.ID]
+		var alloc *quotapool.IntAlloc
+		if cmd.proposal != nil {
+			// INVARIANT: a proposal is consumed (i.e. removed from the proposals map)
+			// the first time it comes up for application. (If the proposal fails due
+			// to an illegal LeaseAppliedIndex, a new proposal might be spawned to
+			// retry but that proposal will be unrelated as far as log application is
+			// concerned).
+			//
+			// INVARIANT: local proposals are not in the proposals map while being
+			// applied, and they never re-enter the proposals map either during or
+			// afterwards.
+			//
+			// (propBuf.{Insert,ReinsertLocked} ignores proposals that have
+			// v2SeenDuringApplicationSet to make this true).
+			if cmd.proposal.v2SeenDuringApplication {
+				log.Fatalf(d.r.AnnotateCtx(context.Background()), "ProposalData seen twice during application: %+v", cmd.proposal)
```

Comment:
Can we make this not be fatal in release builds? Consider `logcrash.ReportOrPanic`.